pubsub/google/cloud/pubsub_v1/subscriber/policy/base.py

```diff
@@ -0,0 +1,392 @@
+
+        # If we have been paused by flow control, check and see if we are
+        # back within our limits.
+        #
+        # In order to not thrash too much, require us to have passed below
+        # the resume threshold (80% by default) of each flow control setting
+        # before restarting.
+        if self._paused and self._load < self.flow_control.resume_threshold:
+            self._paused = False
+            self.open(self._callback)
+
+    def get_initial_request(self, ack_queue=False):
+        """Return the initial request.
+
+        This defines the initial request that must always be sent to Pub/Sub
+        immediately upon opening the subscription.
+
+        Args:
+            ack_queue (bool): Whether to include any acks that were sent
+                while the connection was paused.
+
+        Returns:
+            ~.pubsub_v1.types.StreamingPullRequest: A request suitable
+                for being the first request on the stream (and not suitable
+                for any other purpose).
+
+        .. note::
+            If ``ack_queue`` is set to True, this includes the ack_ids, but
+            also clears the internal set.
+
+            This means that calls to :meth:`get_initial_request` with
+            ``ack_queue`` set to True are not idempotent.
+        """
+        # Any ack IDs that are under lease management and not being acked
+        # need to have their deadline extended immediately.
+        ack_ids = set()
+        lease_ids = self.managed_ack_ids
+        if ack_queue:
+            ack_ids = self._ack_on_resume
+            lease_ids = lease_ids.difference(ack_ids)
+
+        # Put the request together.
+        request = types.StreamingPullRequest(
+            ack_ids=list(ack_ids),
+            modify_deadline_ack_ids=list(lease_ids),
+            modify_deadline_seconds=[self.ack_deadline] * len(lease_ids),
+            stream_ack_deadline_seconds=self.histogram.percentile(99),
+            subscription=self.subscription,
+        )
+
+        # Clear the ack_ids set.
+        # Note: If `ack_queue` is False, this just ends up being a no-op,
+        # since the set is just an empty set.
+        ack_ids.clear()
+
+        # Return the initial request.
+        return request
+
+    def lease(self, ack_id, byte_size):
+        """Add the given ack ID to lease management.
+
+        Args:
+            ack_id (str): The ack ID.
+            byte_size (int): The size of the PubSub message, in bytes.
+        """
+        # Add the ack ID to the set of managed ack IDs, and increment
+        # the size counter.
+        if ack_id not in self.managed_ack_ids:
+            self.managed_ack_ids.add(ack_id)
+            self._bytes += byte_size
+
+        # Sanity check: Do we have too many things in our inventory?
+        # If we do, we need to stop the stream.
+        if self._load >= 1.0:
+            self._paused = True
+            self.close()
```

Comment:
why stop the stream? Would it not be better to leave it open and just not queue any new requests until the load is lower?